In [1]:
import pandas as pd

In [2]:
idle_time_data = pd.read_csv('../data/final_df_points_18_21_class.csv')

In [3]:
idle_time_data.head()

bike_id        lat        lng    dt_start      dt_end           hex_id  \
0     3694  51.323619  12.335887  1514764029  1514764799  881f1a8d91fffff   
1     2508  51.323619  12.335887  1514764035  1514764799  881f1a8d91fffff   
2    41851  51.332415  12.395158  1514764638  1514764799  881f1a1659fffff   
3     3694  51.323619  12.335887  1514764800  1514814103  881f1a8d91fffff   
4     2508  51.323619  12.335887  1514764800  1514851199  881f1a8d91fffff   

   in_zone                   zone_name  end_station_number   temp  ...  month  \
0        1  Pinke Zone Leipzig West 1€                4094  10.80  ...     12   
1        1  Pinke Zone Leipzig West 1€                4094  10.80  ...     12   
2        1   Pinke Zone Leipzig Ost 1€                   0  10.80  ...     12   
3        1  Pinke Zone Leipzig West 1€                4094  10.86  ...      1   
4        1  Pinke Zone Leipzig West 1€                4094  10.86  ...      1   

   day  start_min  over_night  idle_time_next_day  idle_time  idle_time_class  \
0    6       1427           1                 821         12                1   
1    6       1427           1                1439         12                1   
2    6       1437           1                1439          2                1   
3    0          0           0                   0        821                4   
4    0          0           1                 878       1439                4   

   on_station  hex_enc  zone_name_enc  
0           1      320              4  
1           1      320              4  
2           0      133              3  
3           1      320              4  
4           1      320              4  

[5 rows x 25 columns]

In [4]:
TargetVariable = ['idle_time_class']
Predictors = ['bike_id', 'lat', 'lng', 'temp', 'rain', 'snow', 'wind_speed', 'humidity', 'dt_start',
              'hex_enc', 'start_min', 'year', 'month', 'day', 'on_station', 'in_zone', 'zone_name_enc']

In [5]:
X = idle_time_data[Predictors].values
y = idle_time_data[TargetVariable].values

from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, train_size=0.9, shuffle=False)

In [6]:
from sklearn.svm import LinearSVC
from sklearn.naive_bayes import MultinomialNB
from sklearn.multiclass import OneVsRestClassifier
from sklearn.ensemble import RandomForestClassifier, BaggingClassifier, GradientBoostingClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.neighbors import KNeighborsClassifier
from skmultilearn.problem_transform import BinaryRelevance, LabelPowerset

models = []
models.append(('knnc', KNeighborsClassifier()))
models.append(('dtc', DecisionTreeClassifier()))
models.append(('rfc', RandomForestClassifier(n_jobs=-1)))
models.append(('bagc', OneVsRestClassifier(BaggingClassifier(n_jobs=-1))))
models.append(('gbc', OneVsRestClassifier(GradientBoostingClassifier())))
models.append(('bagc', OneVsRestClassifier(LinearSVC(), n_jobs=-1)))
models.append(('gbc', BinaryRelevance(LinearSVC())))
models.append(('nbc', LabelPowerset(LinearSVC())))


In [9]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, confusion_matrix, hamming_loss
ModelsPerformance = {}
results={}

def metricsReport(modelName, test_labels, predictions):
    macro_f1 = f1_score(test_labels, predictions, average='macro')

    micro_f1 = f1_score(test_labels, predictions, average='micro')
    
    hamLoss = hamming_loss(test_labels, predictions)

    res = [macro_f1, micro_f1, hamLoss]
    results[modelName] = res
    print(f"micro_f1: {micro_f1}")
    ModelsPerformance[modelName] = micro_f1

In [10]:
models2 = []
models2.append(('bagc', OneVsRestClassifier(LinearSVC(max_iter= 10000), n_jobs=-1)))
models2.append(('gbc', BinaryRelevance(LinearSVC(max_iter= 10000))))
models2.append(('nbc', LabelPowerset(LinearSVC(max_iter= 10000))))

In [11]:
for name, model in models2:
    print(f"starting with {name}")
    model.fit(X_train, y_train.ravel())
    y_pred = model.predict(X_test)

    metricsReport(name,y_test.ravel(),y_pred)

starting with bagc


/usr/local/Caskroom/miniconda/base/envs/geobike/lib/python3.9/site-packages/sklearn/svm/_base.py:1206: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn(
/usr/local/Caskroom/miniconda/base/envs/geobike/lib/python3.9/site-packages/sklearn/svm/_base.py:1206: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn(
/usr/local/Caskroom/miniconda/base/envs/geobike/lib/python3.9/site-packages/sklearn/svm/_base.py:1206: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn(
/usr/local/Caskroom/miniconda/base/envs/geobike/lib/python3.9/site-packages/sklearn/svm/_base.py:1206: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn(


micro_f1: 0.25244643105168124
starting with gbc


/usr/local/Caskroom/miniconda/base/envs/geobike/lib/python3.9/site-packages/sklearn/svm/_base.py:1206: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn(


TypeError: len() of unsized object

In [ ]:
print(ModelsPerformance)

# res

models = []
models.append(('knnc', KNeighborsClassifier()))
models.append(('dtc', DecisionTreeClassifier()))
models.append(('rfc', RandomForestClassifier(n_jobs=-1)))
models.append(('bagc', OneVsRestClassifier(BaggingClassifier(n_jobs=-1))))
models.append(('gbc', OneVsRestClassifier(GradientBoostingClassifier())))
models.append(('bagc', OneVsRestClassifier(LinearSVC(), n_jobs=-1)))
models.append(('gbc', BinaryRelevance(LinearSVC())))
models.append(('nbc', LabelPowerset(LinearSVC())))


starting with knnc
micro_f1: 0.24607834471844472
starting with dtc
micro_f1: 0.3954653164471718
starting with rfc
micro_f1: 0.48666405714021155
starting with bagc
micro_f1: 0.4442550433307659
starting with gbc
micro_f1: 0.4886717089452042
starting with bagc
micro_f1: 0.25244643105168124